pkgs/development/python-modules/trezor_agent/default.nix

```diff
@@ -10,7 +10,7 @@
 , semver
 , setuptools
 , wheel
-, pinentry
+, pinentry-gtk2
```

Comment:
It's a bit strange that softwares will now be pinned to a specific flavor of pinentry. How can one tell which flavor is to be used?

Reply:
The converse is equally strange. I don't think many users will care about the choice, as long as the choice between graphical and terminal was made appropriately.

> to be used

I agree that `pinentry-gtk2` is too specific here. With other packages, we've used the generic names in the parameter list and added the specific implementations in the `callPackage` arguments.

```nix
callPackage x { pinentry = pkgs.pinentry-gtk2; }
```

Reply:
> The converse is equally strange.

How?

> I don't think many users will care about the choice, as long as the choice between graphical and terminal was made appropriately.

Any time you opt for terminal pinentry you risk to get surprising results once some other graphical tool tries to call your "terminal" tool.

Reply:
> > equally strange.

Having to add `pinentry` separately to your env or having to override it is also strange. Most packages come with their dependencies. Exceptions to this rule do exist, like graphics driver libraries or `nix` itself, which is normally taken from the system environment.

> Any time you opt for terminal pinentry you risk to get surprising results once some other graphical tool tries to call your "terminal" tool.

Yes. It only works for text user interfaces like those built on ncurses; not for commands.
Similarly, a graphical pinentry doesn't work in a purely text environment like the linux console (non-graphical installer, etc)

------

Perhaps we should have a wrapper that only adds pinentry-ncurses if it isn't already on PATH. If the desktop environments provide their own pinentry, this should provide the optimal experience.
 - a graphical pinentry is available when the session is graphical
 - the graphical pinentry will match the desktop's UX
 - pinentry is always available for terminal software
 - small closures
 - fewer rebuilds

On 21.05 I'm getting an ncurses-only pinentry for `pinentry`, so we can change this to

```suggestion
, pinentry-ncurses
```

and follow this up later, afaic. No need for scope creep.

Reply:
> No need for scope creep.

Right, the PR has a benefit of not making `ncurses-gtk` not depend on all dependencies of all other backends. The changes to a couple of minor packages are a distraction: it would have not caused any questions to just leave them to use `{ pinentry = pinentry-ncurses; }` as `${pinentry}/bin/pinentry` was already a symlink to pinentry-ncurses.

Reply:
Right, I think we should just default to `pinentry-curses` for now and let the respective package maintainer decide what to do. My main issue here was that GUI based programs are normally not launched from a terminal, so the ncurses interface cannot be displayed and entering passphrases might not work if the application does not handle the failing pinentry gracefully. That's why using one of the GUI variants is a good idea for UX since they fall back to ncurses gracefully if not launched from within a graphical environment.

Ideally, applications using pinentry would just take the `pinentry` binary from `$PATH` which could contain the appropriate flavor for the user's environment or make it configurable like gpg & gpg-agent do.

Reply:
New version is much better!